File: 1011.在D天内送达包裹的能力/solution.cpp
class Solution {
public:
    int shipWithinDays(vector<int>& weights, int D) {
        int l=*max_element(weights.begin(),weights.end()),r=accumulate(weights.begin(),weights.end(),0);
        while (l<r) {
            int mid=l+(r-l)/2;
            int cur=0,cnt=1;
            for (int i:weights) {
                if (cur+i>mid) {
                    cnt++;
                    cur=i;
                } else {
                    cur+=i;
                }
            }
            if (cnt<=D) {
                r=mid;
            } else {
                l=mid+1;
            }
        }
        return l;
    }
};

File: 1011.在D天内送达包裹的能力/solution.go
func shipWithinDays(weights []int, D int) int {
	l, r := max(weights), sum(weights)
	for l < r {
		mid := l + (r-l)/2
		cur, cnt := 0, 1
		for _, i := range weights {
			if cur+i > mid {
				cur = i
				cnt++
			} else {
				cur += i
			}
		}
		if cnt <= D {
			r = mid
		} else {
			l = mid + 1
		}
	}
	return l
}

func max(weights []int) int {
	mx := 0
	for _, i := range weights {
		if i > mx {
			mx = i
		}
	}
	return mx
}

func sum(weights []int) int {
	sum := 0
	for _, i := range weights {
		sum += i
	}
	return sum
}